File: go/106.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTree1(inorder []int, postorder []int) *TreeNode {
	var m map[int]int = make(map[int]int)
	for i, v := range inorder {
		m[v] = i
	}
	var root = len(postorder) - 1
	return helper(postorder, m, &root, 0, len(postorder)-1)
}

func helper(p []int, m map[int]int, root *int, l int, r int) *TreeNode {
	if r < l {
		return nil
	}
	var pivot int = m[p[*root]]
	var cur TreeNode = TreeNode{Val: p[*root]}
	*root--
	cur.Right = helper(p, m, root, pivot+1, r)
	cur.Left = helper(p, m, root, l, pivot-1)
	return &cur
}

// other's solution, just incredible, fully perfomed the manipution of slice
func buildTree2(inorder []int, postorder []int) *TreeNode {
	if len(postorder) == 0 {
		return nil
	}

	lastIndex := len(postorder) - 1
	root := &TreeNode{Val: postorder[lastIndex]}

	i := search(inorder, postorder[lastIndex])
	root.Left = buildTree2(inorder[:i], postorder[:i])
	root.Right = buildTree2(inorder[i+1:], postorder[i:lastIndex])

	return root
}

func search(data []int, target int) (result int) {
	for ; data[result] != target; result++ {
	}

	return result
}

File: cpp/106.cpp
#include <unordered_map>
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  TreeNode *helper(vector<int> &p, unordered_map<int, int> &m, int &root, int l,
                   int r) {
    if (r < l)
      return nullptr;

    int pivot = m[p[root]];

    TreeNode *cur = new TreeNode(p[root--]);
    cur->right = helper(p, m, root, pivot + 1, r);
    cur->left = helper(p, m, root, l, pivot - 1);

    return cur;
  }
  TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder) {
    unordered_map<int, int> m;
    for (int i = 0; i < inorder.size(); i++)
      m[inorder[i]] = i;
    int lastIndex = postorder.size() - 1;
    return helper(postorder, m, lastIndex, 0, postorder.size() - 1);
  }
};
